File: app/Models/PendingTransfer.php
<?php

namespace App\Models;

use App\Traits\SyncsWithHQ;
use Illuminate\Database\Eloquent\Model;

class PendingTransfer extends Model
{
    use SyncsWithHQ;

    protected $fillable = [
        'reference_number', 'sender_account_id', 'receiver_account_id',
        'sender_user_id', 'receiver_user_id', 'amount', 'currency',
        'exchange_rate', 'description', 'status', 'expires_at', 'accepted_at',
        'declined_at', 'cancelled_at', 'ip_address',
    ];

    protected function casts(): array
    {
        return [
            'amount' => 'decimal:2',
            'expires_at' => 'datetime',
            'accepted_at' => 'datetime',
            'declined_at' => 'datetime',
            'cancelled_at' => 'datetime',
        ];
    }

    // ── Relationships ───────────────────────────────────────────

    public function senderAccount()
    {
        return $this->belongsTo(Account::class, 'sender_account_id');
    }

    public function receiverAccount()
    {
        return $this->belongsTo(Account::class, 'receiver_account_id');
    }

    public function senderUser()
    {
        return $this->belongsTo(User::class, 'sender_user_id');
    }

    public function receiverUser()
    {
        return $this->belongsTo(User::class, 'receiver_user_id');
    }

    // ── Scopes ──────────────────────────────────────────────────

    public function scopePending($query)
    {
        return $query->where('status', 'pending')->where('expires_at', '>', now());
    }

    public function scopeForUser($query, int $userId)
    {
        return $query->where(function ($q) use ($userId) {
            $q->where('sender_user_id', $userId)
              ->orWhere('receiver_user_id', $userId);
        });
    }

    public function scopeIncoming($query, int $userId)
    {
        return $query->where('receiver_user_id', $userId);
    }

    public function scopeOutgoing($query, int $userId)
    {
        return $query->where('sender_user_id', $userId);
    }

    public function scopeExpired($query)
    {
        return $query->where('status', 'pending')->where('expires_at', '<=', now());
    }

    // ── Helpers ─────────────────────────────────────────────────

    public function isPending(): bool
    {
        return $this->status === 'pending' && $this->expires_at->isFuture();
    }

    public function isExpired(): bool
    {
        return $this->status === 'pending' && $this->expires_at->isPast();
    }

    public function canBeAccepted(): bool
    {
        return $this->isPending();
    }

    public function canBeCancelled(): bool
    {
        return $this->isPending();
    }

    public function getTimeRemainingAttribute(): string
    {
        if (!$this->isPending()) return 'Expired';
        return $this->expires_at->diffForHumans(['parts' => 2]);
    }

    /**
     * Generate a unique reference number for pending transfers.
     */
    public static function generateReference(): string
    {
        return 'PTX' . strtoupper(bin2hex(random_bytes(10)));
    }
}

File: mobile/lib/data/models/pending_transfer.dart
class PendingTransfer {
  final int id;
  final String referenceNumber;
  final int senderAccountId;
  final int receiverAccountId;
  final int senderUserId;
  final int receiverUserId;
  final double amount;
  final String currency;
  final double? exchangeRate;
  final String? description;
  final String status;
  final String? expiresAt;
  final String? senderName;
  final String? receiverName;

  const PendingTransfer({
    required this.id,
    required this.referenceNumber,
    required this.senderAccountId,
    required this.receiverAccountId,
    required this.senderUserId,
    required this.receiverUserId,
    required this.amount,
    required this.currency,
    this.exchangeRate,
    this.description,
    required this.status,
    this.expiresAt,
    this.senderName,
    this.receiverName,
  });

  factory PendingTransfer.fromJson(Map<String, dynamic> json) {
    return PendingTransfer(
      id: json['id'] as int,
      referenceNumber: json['reference_number'] as String? ?? '',
      senderAccountId: json['sender_account_id'] as int? ?? 0,
      receiverAccountId: json['receiver_account_id'] as int? ?? 0,
      senderUserId: json['sender_user_id'] as int? ?? 0,
      receiverUserId: json['receiver_user_id'] as int? ?? 0,
      amount: (json['amount'] as num).toDouble(),
      currency: json['currency'] as String? ?? 'USD',
      exchangeRate: json['exchange_rate'] != null
          ? (json['exchange_rate'] as num).toDouble()
          : null,
      description: json['description'] as String?,
      status: json['status'] as String,
      expiresAt: json['expires_at'] as String?,
      senderName: json['sender_name'] as String?,
      receiverName: json['receiver_name'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'reference_number': referenceNumber,
      'sender_account_id': senderAccountId,
      'receiver_account_id': receiverAccountId,
      'sender_user_id': senderUserId,
      'receiver_user_id': receiverUserId,
      'amount': amount,
      'currency': currency,
      'exchange_rate': exchangeRate,
      'description': description,
      'status': status,
      'expires_at': expiresAt,
      'sender_name': senderName,
      'receiver_name': receiverName,
    };
  }
}
